packages/node/src/indexer/mmr.service.ts

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import fs from 'fs';
+import { Injectable, OnApplicationShutdown } from '@nestjs/common';
+import { u8aToHex, u8aEq } from '@polkadot/util';
+import { MMR } from 'merkle-mountain-range';
+import { Sequelize, Op } from 'sequelize';
+import { NodeConfig } from '../configure/NodeConfig';
+import { SubqueryProject } from '../configure/project.model';
+import { getLogger } from '../utils/logger';
+import { delay } from '../utils/promise';
+import { FileBasedDb, keccak256FlyHash } from '../vendor/merkle-mountain-range';
+import { MetadataFactory, MetadataRepo } from './entities/Metadata.entity';
+import { PoiFactory, PoiRepo, ProofOfIndex } from './entities/Poi.entity';
+
+const logger = getLogger('mmr');
+const DEFAULT_WORD_SIZE = 32;
+const DEFAULT_LEAF = Buffer.from('0x000000000000000000000000000000');
+const DEFAULT_FETCH_RANGE = 100;
+
+@Injectable()
+export class MmrService implements OnApplicationShutdown {
+  private isShutdown = false;
+  private blockOffset: number;
+  private metadataRepo: MetadataRepo;
+  private fileBasedMmr: MMR;
+  private poiRepo: PoiRepo;
+  private mmrStartHeight: number;
+
+  constructor(
+    protected nodeConfig: NodeConfig,
+    protected project: SubqueryProject,
+    protected sequelize: Sequelize,
+  ) {}
+
+  onApplicationShutdown(): void {
+    this.isShutdown = true;
+  }
+
+  async init(schema: string): Promise<void> {
+    this.metadataRepo = MetadataFactory(this.sequelize, schema);
+    this.poiRepo = PoiFactory(this.sequelize, schema);
+    this.blockOffset = await this.fetchBlockOffsetFromDb();
+    this.ensureFileBasedMmr(this.nodeConfig.mmrPath);
+    this.mmrStartHeight =
+      (await this.fileBasedMmr.getLeafLength()) + this.blockOffset;
+    logger.info(
+      `file based database MMR start with height ${this.mmrStartHeight}`,
+    );
+  }
+
+  async fetchBlockOffsetFromDb(): Promise<number | null> {
+    const blockOffset = await this.metadataRepo.findOne({
+      where: { key: 'blockOffset' },
+    });
+    if (blockOffset === null) {
+      throw new Error(`Poi service failed to fetch block offset from metadata`);
+    }
+    return Number(blockOffset.value);
+  }
+
+  async appendMmrNode(poiBlock: ProofOfIndex): Promise<void> {
+    let mmrRoot: Uint8Array;
+    const newLeaf = poiBlock.hash;
+    if (newLeaf.length !== DEFAULT_WORD_SIZE) {
+      throw new Error(
+        `Append Mmr failed, input data length should be ${DEFAULT_WORD_SIZE}`,
+      );
+    }
+    const blockLeafHeight = poiBlock.id - this.blockOffset - 1; //leaf index alway -1
```

Comment:
I found the blockOffset you implemented is '0', shouldn't it be '1' ? if we made this change, do we still need `-1` here?

Reply:
I think offset to 0 is make more sense than 1. 

`poiBlock.id - this.blockOffset ` should give us the the total number of blocks in the mmr (aka, leafLength)

`blockOffset` also used in line 46, 

```
    this.mmrStartHeight =
      (await this.fileBasedMmr.getLeafLength()) + this.blockOffset;
```

When mmr service initialization, we want to repeat calculate previous mmr value for validation. Eg, last fileDb mmr block height is 100, we want it repeat generate mmr for block 100 again and compare to value in poi table.
Therefore in this case, it start from 100 + 0 (which is equal to initial start block -1) .

And If we changed the blockOffset to 1 (which is equal to initial start block), it start from 100 + 1 (offset) -1 (to previous block).